### PR TITLE
*: Abstract coprocess result to interface

### PR DESCRIPTION
**Pull request by shenli**
Prepare for mocked xapi

### CODE REVIEW

**Comment by ngaut** [comment]
LGTM

File: executor/new_executor_xapi.go [hunk 1]
```diff
@@ -59,8 +77,8 @@ type NewXSelectIndexExec struct {
 	ctx         context.Context
 	supportDesc bool
 	isMemDB     bool
-	result      *xapi.SelectResult
-	subResult   *xapi.SubResult
+	result      xapi.SelectResult
+	subResult   xapi.PartialResult
```
Comment (coocood) [inline executor/new_executor_xapi.go]: rename to 'partialResult' according to type.

File: xapi/xapi.go [hunk 2]
```diff
@@ -62,17 +86,17 @@ func (r *SelectResult) Next() (subResult *SubResult, err error) {
...
 
 // Close closes SelectResult.
-func (r *SelectResult) Close() error {
+func (r *selectResult) Close() error {
 	return r.resp.Close()
 }
 
 // SubResult represents a subset of select result.
-type SubResult struct {
+type subResult struct {
```
Comment (coocood) [inline xapi/xapi.go]: rename to 'partialResult'?

**Comment by shenli** [comment]
@coocood PTAL

**Comment by shenli** [comment]
@coocood PTAL

File: executor/new_executor_xapi.go [hunk 3]
```diff
@@ -51,17 +51,35 @@ type NewXExecutor interface {
...
+	// Close closes the object.
+	Close() error
+}
+
+func closeAll(objs ...Closeable) error {
+	for _, obj := range objs {
+		if obj != nil {
+			err := obj.Close()
+			if err != nil {
+				return errors.Trace(err)
```
Comment (tiancaiamao) [inline executor/new_executor_xapi.go]: would this cause resource leak?

Comment (tiancaiamao) [inline executor/new_executor_xapi.go]: I think we'd better log the error and continue...
rest LGTM

Comment (tiancaiamao) [inline executor/new_executor_xapi.go]: I think we'd better log the error and continue...
rest LGTM

Comment (shenli) [inline executor/new_executor_xapi.go]: I am not sure how to handle rest closeable objects. May be we could find a better way latter.

File: executor/new_executor_xapi.go [hunk 4]
```diff
@@ -429,10 +429,10 @@ func (e *NewXSelectIndexExec) extractRowsFromTableResult(t table.Table, tblResul
 	return rows, nil
 }
 
-func (e *NewXSelectIndexExec) extractRowsFromSubResult(t table.Table, subResult xapi.PartialResult) ([]*Row, error) {
+func (e *NewXSelectIndexExec) extractRowsFromSubResult(t table.Table, partialResult xapi.PartialResult) ([]*Row, error) {
```
Comment (coocood) [inline executor/new_executor_xapi.go]: Sub -> Partial

File: xapi/xapi.go [hunk 5]
```diff
@@ -96,7 +96,7 @@ func (r *selectResult) Close() error {
 }
 
 // SubResult represents a subset of select result.
```
Comment (coocood) [inline xapi/xapi.go]: SubResult -> partialResult

**Comment by coocood** [comment]
rest LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**